library/core/src/cell.rs

```diff
@@ -430,14 +437,20 @@ impl<T> Cell<T> {
     #[stable(feature = "move_cell", since = "1.17.0")]
     pub fn swap(&self, other: &Self) {
         if ptr::eq(self, other) {
+            // Swapping wouldn't change anything.
             return;
         }
+        if !is_nonoverlapping(self, other, 1) {
+            // See <https://github.com/rust-lang/rust/issues/80778> for why we need to stop here.
+            panic!("`Cell::swap` on overlapping non-identical `Cell`s");
+        }
         // SAFETY: This can be risky if called from separate threads, but `Cell`
         // is `!Sync` so this won't happen. This also won't invalidate any
         // pointers since `Cell` makes sure nothing else will be pointing into
-        // either of these `Cell`s.
+        // either of these `Cell`s. We also excluded shenanigans like partially overlapping `Cell`s,
+        // so `swap` will just properly copy two full values of type `T` back and forth.
         unsafe {
-            ptr::swap(self.value.get(), other.value.get());
+            ptr::swap_nonoverlapping(self.value.get(), other.value.get(), 1);
```

Comment:
Pondering: would it be reasonable for this to `mem::swap(&mut*self.value.get(), &mut*other.value.get())` now that they're non-overlapping?  Or is it important that it stays in pointers here?

r=me either way from a code perspective.

Reply:
We could use `mem::swap`, yeah. What would be the advantage of that?

Reply:
Probably doesn't make a substantial difference.  I think of it as simpler since it doesn't need the loop (usually), but I suppose it'll still instantiate the `ptr::swap_nonoverlapping` code anyway.